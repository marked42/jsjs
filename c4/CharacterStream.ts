export class CharacterStream {
  static EOF = -1
  private index = 0
  constructor(private input: string) {}

  peek() {
    const code = this.input.charCodeAt(this.index)
    if (Number.isNaN(code)) {
      return CharacterStream.EOF
    }

    return code
  }

  next() {
    const code = this.peek()
    if (code !== CharacterStream.EOF) {
      this.index++
    }

    return code
  }

  prev() {
    this.index -= 1
  }
}
