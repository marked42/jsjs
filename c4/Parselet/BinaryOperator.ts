import { PrecedentialOperator, OperatorAssociativity } from './Operator'

/**
 * 二元操作符
 */

export class BinaryOperator extends PrecedentialOperator {
  constructor(precedence: number, associativity: OperatorAssociativity) {
    super(precedence, {
      hasPrecedingOperand: true,
      hasFollowingOperand: true,
      isFirstOperator: true,
      isLastOperator: true,
      expressionStartWithOperand: true,
      associativity,
    })
  }
}
