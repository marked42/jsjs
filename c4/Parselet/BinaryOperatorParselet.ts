import type { InfixParselet } from './Parselet'
import type { ParseletParser } from './ParseletParser'
import { BinaryExpression, Expression } from '../AST'
import type { Token } from '../Token'
import { BinaryOperator } from './Operator'

export class BinaryOperatorParselet implements InfixParselet {
  constructor(
    private op: BinaryOperator,
    private compose?: (
      left: Expression,
      right: Expression,
      token: Token
    ) => Expression
  ) {}

  parse(parser: ParseletParser, result: Expression, token: Token) {
    const right = parser.expression(this.op.rightBindingPower())

    const defaultCompose = (
      left: Expression,
      right: Expression,
      token: Token
      // @ts-ignore TODO:
    ) => new BinaryExpression(left, right, token.source as any)
    const compose = this.compose || defaultCompose
    return compose(result, right, token)
  }

  getPrecedence() {
    return this.op.leftBindingPower()
  }
}
