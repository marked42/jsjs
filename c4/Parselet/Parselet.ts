import { ParseletParser } from './ParseletParser'
import { Expression } from '../AST'
import { Token } from '../Token'

export interface PrefixParselet {
  parse(parser: ParseletParser, token: Token): Expression
}

export interface InfixParselet {
  parse(parser: ParseletParser, result: Expression, token: Token): Expression
  getPrecedence(): number
}
