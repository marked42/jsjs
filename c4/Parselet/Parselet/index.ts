export * from './AtomParselet'
export * from './BinaryOperatorParselet'
export * from './NonFirstOperatorParselet'
export * from './OperatorInfixParselet'
export * from './Parselet'
export * from './PostfixOperatorParselet'
export * from './PrefixOperatorParselet'
