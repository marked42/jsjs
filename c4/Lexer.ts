import { CharacterStream } from './CharacterStream'
import { Token, TokenType, Identifier, StringLiteral } from './Token'
import * as Errors from './Errors'

export class Lexer {
  private codes: number[] = []
  private charCode: number = -1
  constructor(private stream: CharacterStream) {}

  getChar() {
    this.charCode = this.stream.next()
    if (this.charCode !== CharacterStream.EOF) {
      this.codes.push(this.charCode)
    }
    return this.charCode
  }

  ungetChar() {
    this.codes.splice(this.codes.length - 1, 1)
    this.stream.prev()
    this.charCode = this.codes[this.codes.length - 1]
  }

  resetCharCodes() {
    this.codes = []
    this.charCode = -1
  }

  next(): Token {
    this.getChar()
    // StringLiteral
    if (charCodeIs(this.charCode, '"')) {
      const value = []

      while (true) {
        this.getChar()

        if (charCodeIs(this.charCode, '\\')) {
          this.getChar()
          if (charCodeIs(this.charCode, '"')) {
            value.push('"'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, "'")) {
            value.push("'".charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'b')) {
            value.push('\b'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'f')) {
            value.push('\f'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'n')) {
            value.push('\n'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'r')) {
            value.push('\r'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 't')) {
            value.push('\t'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'v')) {
            value.push('\v'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, '\\')) {
            value.push('\\'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, '\r')) {
            value.push('\r'.charCodeAt(0))
            this.getChar()
            if (charCodeIs(this.charCode, '\n')) {
              value.push('\n'.charCodeAt(0))
            } else {
              this.ungetChar()
            }
          } else if (charCodeIs(this.charCode, '\n')) {
            value.push('\n'.charCodeAt(0))
          } else {
            console.error(
              'invalid escape sequence \\' + String.fromCharCode(this.charCode)
            )
            value.push(this.charCode)
          }
        } else if (
          charCodeIs(this.charCode, '\r') ||
          charCodeIs(this.charCode, '\n')
        ) {
          throw Errors.StringLiteralUnescapedNewline
        } else if (charCodeIs(this.charCode, '"')) {
          const source = String.fromCharCode(...this.codes)
          this.resetCharCodes()
          return {
            type: TokenType.StringLiteral,
            source,
            value: String.fromCharCode(...value),
          } as StringLiteral
        } else if (this.charCode === CharacterStream.EOF) {
          throw Errors.StringLiteralEndEarly
        } else {
          value.push(this.charCode)
        }
      }
      // NumericLiteral
    } else if (isDigit(this.charCode)) {
      // LineComment
    } else if (charCodeIs(this.charCode, '/')) {
      this.stream.next()

      if (charCodeIs(this.stream.peek(), '/')) {
        this.stream.next()
        const codes = []
        while (true) {
          const code = this.stream.peek()
          if (!charCodeIs(code, '\r') && !charCodeIs(code, '\n')) {
            codes.push(code)
            this.stream.next()
          } else {
            return {
              type: TokenType.LineComment,
              source: String.fromCharCode(...codes),
            }
          }
        }
      } else {
        return {
          type: TokenType.Operator,
          source: '/',
        }
      }
      // Whitespace
    } else if (isWhitespace(this.charCode)) {
      const codes = [this.stream.next()]
      while (isWhitespace(this.stream.peek())) {
        codes.push(this.stream.next())
      }

      return {
        type: TokenType.Whitespace,
        source: String.fromCharCode(...codes),
      }
      // Newline
    } else if (charCodeIs(this.charCode, '\r')) {
      this.stream.next()
      if (charCodeIs(this.stream.peek(), '\n')) {
        this.stream.next()
        return {
          type: TokenType.Newline,
          source: '\r\n',
        }
      } else {
        return {
          type: TokenType.Newline,
          source: '\r',
        }
      }
    } else if (charCodeIs(this.charCode, '\n')) {
      this.stream.next()
      return {
        type: TokenType.Whitespace,
        source: '\n',
      }
    } else if (isIdentifierStart(this.charCode)) {
      const codes = [this.stream.next()]

      while (true) {
        const code = this.stream.peek()
        if (isIdentifierContinue(code)) {
          codes.push(code)
          this.stream.next()
          continue
        }

        const source = String.fromCharCode(...codes)
        return {
          type: TokenType.Identifier,
          source,
          name: source,
        } as Identifier
      }
    }

    throw new Error('Unexpected token ' + String.fromCharCode(this.charCode))
  }
}

function charCodeIs(code: number, text: string) {
  if (text.length !== 1) {
    throw new Error('text must be single character, got ' + text)
  }
  return code === text.charCodeAt(0)
}

function isDigit(code: number) {
  return code >= '0'.charCodeAt(0) && code <= '9'.charCodeAt(0)
}

function isWhitespace(code: number) {
  const ws = [' ', '\t'].map((c) => c.charCodeAt(0))

  return ws.includes(code)
}

function isLetter(code: number) {
  const isLower = code >= 'a'.charCodeAt(0) && code <= 'z'.charCodeAt(0)
  const isUpper = code >= 'A'.charCodeAt(0) && code <= 'Z'.charCodeAt(0)

  return isLower || isUpper
}

function isIdentifierStart(code: number) {
  return isLetter(code) || charCodeIs(code, '_')
}

function isIdentifierContinue(code: number) {
  return isLetter(code) || charCodeIs(code, '_') || isDigit(code)
}
