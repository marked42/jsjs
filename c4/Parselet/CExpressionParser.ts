import { ParseletParser } from './ParseletParser'
import { Lexer } from '../Lexer'
import {
  NumberValueToken,
  StringNameToken,
  StringValueToken,
  Token,
  TokenType,
} from '../Token'
import {
  PrefixOperatorParselet,
  BinaryOperatorParselet,
  PostfixOperatorParselet,
  NonFirstOperatorParselet,
  AtomParselet,
  OperatorInfixParselet,
} from './Parselet'
import {
  ConditionalExpression,
  Expression,
  Identifier,
  MemberExpression,
  NumericLiteral,
  StringLiteral,
  CallExpression,
  SequenceExpression,
  AssignmentExpression,
  LogicalExpression,
} from '../AST'
import {
  PrecedentialOperator,
  PrefixOperator,
  PostfixOperator,
  BinaryOperator,
  MiddleOperator,
  LastAssociativeOperator,
  OperatorPrecedence,
} from './Operator'

const enum ExpressionPrecedence {
  PlusMinus = 1,
  StarDiv = 2,
  StarStar = 3,
  Conditional = 3,
  MemberExpression = 14,
  IndexExpression = 14,
  PrefixMinusPlus = 14,
  Increment = 15,
  Decrement = 15,
  Parenthesis = 15,
  CallExpression = 15,
}

const enum Precedence {
  /**
   * a++, a--
   */
  PostfixIncrementDecrement = 1,
  /**
   * fn(1)
   */
  FunctionCall = 1,
  /**
   * a[1]
   */
  ArraySubscripting = 1,
  /**
   * a.b  a->b
   */
  MemberAccess = 1,
  PointerMemberAccess = 1,
  /**
   * (type) { list }
   */
  CompoundLiteral = 1,

  /**
   * ++a, --a
   */
  PrefixIncrementDecrement = 2,
  /**
   * -a, +a
   */
  UnaryPlusMinus = 2,
  /**
   * !a
   */
  LogicalNot = 2,
  /**
   * ~a
   */
  BitwiseNot = 2,

  /**
   * (type)
   */
  Cast = 2,

  /**
   * *
   */
  Dereference = 2,

  /** & */
  AddressOf = 2,

  /** sizeof */
  Sizeof = 2,

  /** _Alignof */
  Alignof = 2,

  /** * / % */
  MultiplicationDivisionRemainder = 3,

  /** + - */
  AdditionSubtraction = 4,

  /** << >> */
  BitwiseShift = 5,

  /** < <= > >= */
  LtLeGtGe = 6,

  /**
   * == !=
   */
  EqNq = 7,

  /** & */
  BitwiseAnd = 8,

  /** ^ */
  BitwiseXor = 9,

  /** | */
  BitwiseOr = 10,

  /** && */
  LogicalAnd = 11,

  /** || */
  LogicalOr = 12,

  Conditional = 13,

  Assignment = 14,

  Comma = 15,
}

export class CParser extends ParseletParser {
  constructor(lexer: Lexer) {
    super(lexer)

    this.registerAtoms()
    // this.registerPrefixOperators()
    // this.registerPostfixOperators()
    this.registerBinaryOperators()
    this.registerSequenceExpression()
    // this.registerParenthesis()
    // this.registerIndexOperators()
    this.registerConditionalOperators()
    // this.registerDot()
    // this.registerCallExpression()
  }

  registerAtoms() {
    this.registerPrefixParselet(
      TokenType.Identifier,
      new AtomParselet<StringNameToken>((token) => {
        return new Identifier(token.name)
      })
    )
    this.registerPrefixParselet(
      TokenType.NumericLiteral,
      new AtomParselet<NumberValueToken>(
        (token) => new NumericLiteral(token.value)
      )
    )
    this.registerPrefixParselet(
      TokenType.StringLiteral,
      new AtomParselet<StringValueToken>(
        (token) => new StringLiteral(token.value)
      )
    )
  }

  registerDot() {
    this.registerInfixParselet(
      TokenType.Dot,
      new BinaryOperatorParselet(
        new BinaryOperator(ExpressionPrecedence.MemberExpression, 'left'),
        (left: Expression, right: Expression, token: Token) =>
          new MemberExpression(left, right, false)
      )
    )
  }

  registerIndexOperators() {
    // 索引表达式a[b]
    const leftSquare = new PrecedentialOperator(
      ExpressionPrecedence.IndexExpression,
      {
        hasPrecedingOperand: true,
        hasFollowingOperand: true,
        isFirstOperator: true,
        isLastOperator: false,
        expressionStartWithOperand: true,
      }
    )
    const rightSquare = new PrecedentialOperator(
      ExpressionPrecedence.IndexExpression,
      {
        hasPrecedingOperand: true,
        hasFollowingOperand: false,
        isFirstOperator: false,
        isLastOperator: true,
        expressionStartWithOperand: true,
      }
    )

    this.registerInfixParselet(
      TokenType.LeftSquare,
      new (class extends OperatorInfixParselet {
        parse(parser: ParseletParser, result: Expression, token: Token) {
          const property = parser.expression(leftSquare.rightBindingPower())
          parser.consume(TokenType.RightSquare)

          return new MemberExpression(result, property, true)
        }
      })(leftSquare)
    )

    this.registerInfixParselet(
      TokenType.RightSquare,
      new NonFirstOperatorParselet(rightSquare)
    )
  }

  registerParenthesis() {
    const leftParen = new PrecedentialOperator(
      ExpressionPrecedence.Parenthesis,
      {
        hasPrecedingOperand: false,
        hasFollowingOperand: true,
        isFirstOperator: true,
        isLastOperator: false,
        expressionStartWithOperand: false,
      }
    )
    const rightParen = new PrecedentialOperator(
      ExpressionPrecedence.Parenthesis,
      {
        hasPrecedingOperand: true,
        hasFollowingOperand: false,
        isFirstOperator: false,
        isLastOperator: true,
        expressionStartWithOperand: false,
      }
    )
    // 括号表达式
    this.registerPrefixParselet(TokenType.LeftParen, {
      parse(parser: ParseletParser, token: Token) {
        const expr = parser.expression(leftParen.rightBindingPower())

        expr.parenthesized = true
        parser.consume(TokenType.RightParen)

        return expr
      },
    })
    this.registerInfixParselet(
      TokenType.RightParen,
      new NonFirstOperatorParselet(rightParen)
    )
  }

  registerCallExpression() {
    const leftParen = new PrecedentialOperator(
      ExpressionPrecedence.CallExpression,
      {
        hasPrecedingOperand: true,
        hasFollowingOperand: true,
        isFirstOperator: true,
        isLastOperator: false,
        expressionStartWithOperand: true,
      }
    )
    const comma = new MiddleOperator(ExpressionPrecedence.CallExpression)
    const rightParen = new PrecedentialOperator(
      ExpressionPrecedence.CallExpression,
      {
        hasPrecedingOperand: true,
        hasFollowingOperand: false,
        isFirstOperator: false,
        isLastOperator: true,
        expressionStartWithOperand: false,
      }
    )

    this.registerInfixParselet(
      TokenType.LeftParen,
      new (class extends OperatorInfixParselet {
        parse(parser: ParseletParser, result: Expression, token: Token) {
          const params = [parser.expression(leftParen.rightBindingPower())]

          while (parser.lookahead(0).type !== TokenType.RightParen) {
            parser.consume(TokenType.Comma)
            params.push(parser.expression(leftParen.rightBindingPower()))
          }
          parser.consume(TokenType.RightParen)

          return new CallExpression(result, params)
        }
      })(leftParen)
    )
    this.registerInfixParselet(
      TokenType.Comma,
      new NonFirstOperatorParselet(comma)
    )
    // 和括号表达式中的‘)'重复了，但是不影响逻辑逻辑正确性
    this.registerInfixParselet(
      TokenType.RightParen,
      new NonFirstOperatorParselet(rightParen)
    )
  }

  registerConditionalOperators() {
    const question = new PrecedentialOperator(Precedence.Conditional, {
      hasPrecedingOperand: true,
      hasFollowingOperand: true,
      isFirstOperator: true,
      isLastOperator: false,
      expressionStartWithOperand: true,
    })
    const colon = new LastAssociativeOperator(Precedence.Conditional, 'right')

    this.registerInfixParselet(
      TokenType.Question,
      new (class extends OperatorInfixParselet {
        parse(
          parser: ParseletParser,
          result: Expression,
          token: Token
        ): Expression {
          const consequent = parser.expression(this.leftBindingPower())
          parser.consume(TokenType.Colon)
          const alternate = parser.expression(colon.rightBindingPower())

          return new ConditionalExpression(result, consequent, alternate)
        }
      })(question)
    )

    this.registerInfixParselet(
      TokenType.Colon,
      new NonFirstOperatorParselet(colon)
    )
  }

  registerPrefixOperators() {
    const operators = [
      [
        TokenType.Minus,
        new PrefixOperator(ExpressionPrecedence.PrefixMinusPlus),
      ],
      [
        TokenType.Plus,
        new PrefixOperator(ExpressionPrecedence.PrefixMinusPlus),
      ],
    ] as const
    operators.forEach(([type, op]) => {
      this.registerPrefixParselet(type, new PrefixOperatorParselet(op))
    })
  }

  registerPostfixOperators() {
    const operators = [
      [TokenType.Increment, new PostfixOperator(15)],
      [TokenType.Decrement, new PostfixOperator(15)],
    ] as const
    operators.forEach(([type, op]) => {
      this.registerInfixParselet(type, new PostfixOperatorParselet(op))
    })
  }

  registerBinaryOperators() {
    const operators = [
      TokenType.Assign,
      TokenType.PlusAssign,
      TokenType.MinusAssign,
      TokenType.StarAssign,
      TokenType.DivAssign,
      TokenType.ModulusAssign,
      TokenType.AndAssign,
      TokenType.OrAssign,
      TokenType.XorAssign,
      TokenType.LeftShiftAssign,
      TokenType.RightShiftAssign,
    ]
    const op = new BinaryOperator(Precedence.Assignment, 'right')
    operators.forEach((type) => {
      this.registerInfixParselet(
        type,
        new BinaryOperatorParselet(op, (left, right, token) => {
          // @ts-ignore TODO:
          return new AssignmentExpression(left, right, token.source)
        })
      )
    })

    const logicalOperators = [
      [TokenType.OrOr, new BinaryOperator(Precedence.LogicalOr, 'left')],
      [TokenType.AndAnd, new BinaryOperator(Precedence.LogicalAnd, 'left')],
    ] as const
    logicalOperators.forEach(([type, op]) => {
      this.registerInfixParselet(
        type,
        new BinaryOperatorParselet(op, (left, right, token) => {
          // @ts-ignore TODO:
          return new LogicalExpression(left, right, token.source)
        })
      )
    })
  }

  // 两种构造方法
  // 1. 左结合的二元运算符
  registerSequenceExpression() {
    const precedence = OperatorPrecedence.of(Precedence.Comma)
    this.registerInfixParselet(TokenType.Comma, {
      parse(parser: ParseletParser, result: Expression, token: Token) {
        const params = [result]

        while (true) {
          const result = parser.expression(
            precedence.higherPrecedenceByOneLevel()
          )
          params.push(result)
          if (parser.lookahead(0).type !== TokenType.Comma) {
            break
          }
          parser.consume(TokenType.Comma)
        }
        return new SequenceExpression(params)
      },
      leftBindingPower() {
        return precedence
      },
    })
  }

  // 2. 右结合的二元运算符，对结果进行处理成平行的ast
  // registerSequenceExpression() {
  //   const precedence = OperatorPrecedence.of(Precedence.Comma)
  //   this.registerInfixParselet(TokenType.Comma, {
  //     parse(parser: ParseletParser, result: Expression, token: Token) {
  //       const params = [result]

  //       while (true) {
  //         const result = parser.expression(precedence)
  //         if (result instanceof SequenceExpression) {
  //           params.splice(0, 0, ...result.expressions)
  //         } else {
  //           params.push(result)
  //         }
  //         if (parser.lookahead(0).type !== TokenType.Comma) {
  //           break
  //         }
  //         parser.consume(TokenType.Comma)
  //       }
  //       return new SequenceExpression(params)
  //     },
  //     leftBindingPower() {
  //       return precedence
  //     },
  //   })
  // }
}
