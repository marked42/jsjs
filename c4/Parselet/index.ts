import { ParseletParser } from './ParseletParser'
import { Lexer } from '../Lexer'
import { Token, TokenType } from '../Token'
import { IdentifierParselet } from './IdentifierParselet'
import { NumericLiteralParselet } from './NumericLiteralParselet'
import { PrefixOperatorParselet } from './PrefixOperatorParselet'
import { BinaryOperatorParselet } from './BinaryOperatorParselet'
import { OperatorAssociativity } from '../Operators'
import { PostfixOperatorParselet } from './PostfixOperatorParselet'
import { LeftParenParselet } from './LeftParenParselet'
import { EndTokenParselet } from './EndTokenParselet'
import { LeftSquareParselet } from './IndexOperatorParselet'
import { Expression, MemberExpression } from '../AST'

export class Parser extends ParseletParser {
  constructor(lexer: Lexer) {
    super(lexer)

    this.registerPrefixParselet(TokenType.Identifier, new IdentifierParselet())
    this.registerPrefixParselet(
      TokenType.NumericLiteral,
      new NumericLiteralParselet()
    )

    // 括号表达式
    this.registerPrefixParselet(TokenType.LeftParen, new LeftParenParselet())
    this.registerInfixParselet(TokenType.RightParen, new EndTokenParselet())

    // 索引表达式a[b]
    this.registerInfixParselet(TokenType.LeftSquare, new LeftSquareParselet(14))
    this.registerInfixParselet(TokenType.RightSquare, new EndTokenParselet())

    this.registerPostfixOperators([
      [TokenType.Increment, 15],
      [TokenType.Decrement, 15],
    ])

    this.registerPrefixOperators([
      [TokenType.Minus, 14],
      [TokenType.Plus, 14],
    ])

    this.registerBinaryOperators([
      [TokenType.Plus, 1, 'left'],
      [TokenType.Minus, 1, 'left'],
      [TokenType.Star, 2, 'left'],
      [TokenType.Div, 2, 'left'],
      [TokenType.StarStar, 3, 'right'],
      // [TokenType.Dot, 14, 'left'],
      // TODO:
      //   [TokenType.LeftBracket, 3, 'right'],
      //   [TokenType.Question, 3, 'right'],
    ])

    this.registerInfixParselet(
      TokenType.Dot,
      new (class extends BinaryOperatorParselet {
        composeExpression(left: Expression, right: Expression, token: Token) {
          return new MemberExpression(left, right, false)
        }
      })(14, 'left')
    )
  }

  registerPrefixOperators(prefixOperators: Array<[TokenType, number]>) {
    prefixOperators.forEach(([type, precedence]) => {
      this.registerPrefixParselet(type, new PrefixOperatorParselet(precedence))
    })
  }

  registerPostfixOperators(postfixOperators: Array<[TokenType, number]>) {
    postfixOperators.forEach(([type, precedence]) => {
      this.registerInfixParselet(type, new PostfixOperatorParselet(precedence))
    })
  }

  registerBinaryOperators(
    binaryOperators: Array<[TokenType, number, OperatorAssociativity]>
  ) {
    binaryOperators.forEach(([type, precedence, assoc]) => {
      this.registerInfixParselet(
        type,
        new BinaryOperatorParselet(precedence, assoc)
      )
    })
  }
}
