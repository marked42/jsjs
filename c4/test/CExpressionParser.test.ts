import { CParser } from '../Parselet/CParser'
import { Lexer } from '../Lexer'
import { CharacterStream } from '../CharacterStream'

function parserForInput(input: string) {
  const stream = new CharacterStream(input)
  const lexer = new Lexer(stream)
  const parser = new CParser(lexer)

  return parser
}

describe('comma expression', () => {
  it('should parse comma expression priority 15', () => {
    const input = 'a, b, 1'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })
})

describe('assignment expression', () => {
  it('should parse assignment expression', () => {
    const input = 'a = b += c -= d *= e /= f &= g |= h ^= i'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })
})

describe('conditional expression', () => {
  it('should parse conditional expression', () => {
    const input = 'a ? b : d ? e : f'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse conditional expression with correct precedence', () => {
    const input = 'a = 1 ? b : c = d'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })
})

describe('logical expression', () => {
  it('should parse logical expression', () => {
    const input = 'a || b && c'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })
})

describe('binary expression', () => {
  it('should parse bitwise binary expression', () => {
    const input = 'a | b ^ c & d'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse relational binary expression', () => {
    const input = 'a == b != c < d <= f >= e > g'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse bitwise shift binary expression', () => {
    const input = 'a << b >> c'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse arithmetic binary expression', () => {
    const input = 'a + b - c * d / e % f'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse member expression', () => {
    const input = '2*a.b->c+1'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })
})

describe('prefix operators', () => {
  it('should parse prefix expression', () => {
    const input = '++a, --b, +c, -d, !e, ~f, *g, &h, sizeof i'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })
})

describe('postfix operators', () => {
  it('should parse postfix expression', () => {
    const input = 'a--, b++'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse postfix expression with correct precedence', () => {
    const input = '++a--'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })
})

describe('array subscripting operator', () => {
  it('should parse []', () => {
    const input = 'a + b[c]'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })
})

describe('function call', () => {
  it('should parse function call', () => {
    const input = 'a(b, c)'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })
})

describe('EnumDeclaration', () => {
  it('should parse enum declaration', () => {
    const input = `
    enum Dir {
      Left = 1,
      Right,
      Down
    }
    `
    const parser = parserForInput(input)
    expect(parser.program()).toMatchSnapshot()
  })
})
