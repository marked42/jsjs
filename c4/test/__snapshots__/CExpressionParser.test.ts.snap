// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assignment expression should parse assignment expression 1`] = `
AssignmentExpression {
  "left": Identifier {
    "name": "a",
    "parenthesized": false,
    "type": "Identifier",
  },
  "op": "=",
  "parenthesized": false,
  "right": AssignmentExpression {
    "left": Identifier {
      "name": "b",
      "parenthesized": false,
      "type": "Identifier",
    },
    "op": "+=",
    "parenthesized": false,
    "right": AssignmentExpression {
      "left": Identifier {
        "name": "c",
        "parenthesized": false,
        "type": "Identifier",
      },
      "op": "-=",
      "parenthesized": false,
      "right": AssignmentExpression {
        "left": Identifier {
          "name": "d",
          "parenthesized": false,
          "type": "Identifier",
        },
        "op": "*=",
        "parenthesized": false,
        "right": AssignmentExpression {
          "left": Identifier {
            "name": "e",
            "parenthesized": false,
            "type": "Identifier",
          },
          "op": "/=",
          "parenthesized": false,
          "right": AssignmentExpression {
            "left": Identifier {
              "name": "f",
              "parenthesized": false,
              "type": "Identifier",
            },
            "op": "&=",
            "parenthesized": false,
            "right": AssignmentExpression {
              "left": Identifier {
                "name": "g",
                "parenthesized": false,
                "type": "Identifier",
              },
              "op": "|=",
              "parenthesized": false,
              "right": AssignmentExpression {
                "left": Identifier {
                  "name": "h",
                  "parenthesized": false,
                  "type": "Identifier",
                },
                "op": "^=",
                "parenthesized": false,
                "right": Identifier {
                  "name": "i",
                  "parenthesized": false,
                  "type": "Identifier",
                },
                "type": "AssignmentExpression",
              },
              "type": "AssignmentExpression",
            },
            "type": "AssignmentExpression",
          },
          "type": "AssignmentExpression",
        },
        "type": "AssignmentExpression",
      },
      "type": "AssignmentExpression",
    },
    "type": "AssignmentExpression",
  },
  "type": "AssignmentExpression",
}
`;

exports[`comma expression should parse comma expression priority 15 1`] = `
SequenceExpression {
  "expressions": Array [
    Identifier {
      "name": "a",
      "parenthesized": false,
      "type": "Identifier",
    },
    Identifier {
      "name": "b",
      "parenthesized": false,
      "type": "Identifier",
    },
    NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 1,
    },
  ],
  "parenthesized": false,
  "type": "SequenceExpression",
}
`;

exports[`conditional expression should parse conditional expression 1`] = `
ConditionalExpression {
  "alternate": ConditionalExpression {
    "alternate": Identifier {
      "name": "f",
      "parenthesized": false,
      "type": "Identifier",
    },
    "consequent": Identifier {
      "name": "e",
      "parenthesized": false,
      "type": "Identifier",
    },
    "parenthesized": false,
    "test": Identifier {
      "name": "d",
      "parenthesized": false,
      "type": "Identifier",
    },
    "type": "ConditionalExpression",
  },
  "consequent": Identifier {
    "name": "b",
    "parenthesized": false,
    "type": "Identifier",
  },
  "parenthesized": false,
  "test": Identifier {
    "name": "a",
    "parenthesized": false,
    "type": "Identifier",
  },
  "type": "ConditionalExpression",
}
`;

exports[`conditional expression should parse conditional expression with correct precedence 1`] = `
AssignmentExpression {
  "left": Identifier {
    "name": "a",
    "parenthesized": false,
    "type": "Identifier",
  },
  "op": "=",
  "parenthesized": false,
  "right": AssignmentExpression {
    "left": ConditionalExpression {
      "alternate": Identifier {
        "name": "c",
        "parenthesized": false,
        "type": "Identifier",
      },
      "consequent": Identifier {
        "name": "b",
        "parenthesized": false,
        "type": "Identifier",
      },
      "parenthesized": false,
      "test": NumericLiteral {
        "parenthesized": false,
        "type": "NumericLiteral",
        "value": 1,
      },
      "type": "ConditionalExpression",
    },
    "op": "=",
    "parenthesized": false,
    "right": Identifier {
      "name": "d",
      "parenthesized": false,
      "type": "Identifier",
    },
    "type": "AssignmentExpression",
  },
  "type": "AssignmentExpression",
}
`;
